### PR TITLE
WIP: Webpack v4

### PR DESCRIPTION
**Pull request by dmbch**
In this PR and branch, we are preparing the forthcoming upgrade to [Webpack v4](https://medium.com/webpack/webpack-4-beta-try-it-today-6b1d27d7d7e2).

- [X] upgrade webpack to v4
- [x] update configs to basically work
- [x] re-enable extract-text-webpack-plugin
- [x] review and thoroughly revise configs
- [x] fix chunking (CommonsChunkPlugin is [deprecated](https://gist.github.com/sokra/1522d586b8e5c0f5072d7565c2bee693))
- [x] get rid of `manifest.js` mechanism (config, react, template...)
- [x] investigate `sideEffects: false`

### CODE REVIEW

**Comment by dmbch** [comment]
Re `sideEffects: false`: this is an ESM-only feature - we ought to enable this as soon as `hops-{react, redux,graphql}` migrated to ESNext.

**Comment by ZauberNerd** [comment]
Waiting for one of the following issues/PRs to be resolved:
- https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/701
- https://github.com/webpack/webpack/pull/6448

We could also investigate switching to `extract-css-chunks-webpack-plugin` once this PR is merged:
- https://github.com/faceyspacey/extract-css-chunks-webpack-plugin/pull/61

**Comment by dmbch** [comment]
@ZauberNerd since we discussed making this upgrade non-breaking, I checked what we broke, here. As far as I can tell, these are the two breaking changes introduced in this PR. We could re-introduce `startServer` (paging @robin-drexler), but honestly, removing it is quite an easy fix.

The first one, however, is the bigger issue. Reverting/mitigating that one would essentially make us re-implement WDS...

56315a3: refactor(build): get rid of WDS, use Express
```
BREAKING CHANGE: Since we got rid of webpack dev server, its config is
no longer supported. All existing hops features (including SSL) are
still supported, though.
```

d11c9d5: refactor(express): convert exports to getters
```
BREAKING CHANGE: remove legacy startServer export
```

What do you think?

File: packages/build-config/index.js [hunk 1]
```diff
@@ -2,7 +2,16 @@
 
 var hopsConfig = require('hops-config');
 
-module.exports = Object.freeze({
+module.exports = exports = {
+  get build() {
```
Comment (ZauberNerd) [inline packages/build-config/index.js]: :+1:

File: packages/build-config/loaders/config.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ function getConfig(config) {
           .reduce(function(result, key) {
             result[key] = (function shorten(item) {
               if (typeof item === 'string') {
-                return item.replace(new RegExp(config.appDir), '.');
+                return path.relative(config.appDir, item);
```
Comment (ZauberNerd) [inline packages/build-config/loaders/config.js]: :heart:

**Comment by ZauberNerd** [comment]
LGTM, Although I'd prefer to split this into fixes / features / refactoring that we can deploy with the current Hops@10 and the breaking changes / webpack 4 integration that we'll release as a prerelease.

**Comment by dmbch** [comment]
@ZauberNerd agreed. Could you help with that?

And how do you think we should deal with the DevServer stuff? If we treat it as non-breaking, we could merge it before the Webpack stuff...

**Comment by dmbch** [comment]
@ZauberNerd should we not work on #417 in lieu of this PR, i.e. close this..?

**Pull request closed without merging**